### ISSUE TITLE
CCIDesq One-way integration

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#1680]
### New Feature / Enhancement
Our bookstore is implementing a website redevelopment using CCIDesq eCommerce CMS.  I want to write the integration to OSPOS to send CCIDesq changes to inventory (add, remove, update) when they happen on the OSPOS side.  This would be a one-way integration for now.  Changes are made via RESTful API calls to CCIDesq.  Here are the parameters that I have so far:
- This would be an optional feature turned off by default in OSPOS (similar to the way mailchip is optional).
- When turned on, and configured API calls would be triggered on sale of an item, receiving, return, excel import of items and any other inventory change that I haven't thought of yet.
- Currently there are no plans for two-way data exchange as that would require REST being implemented in OSPOS and my vision of the implementation is that OSPOS, not the CMS would be the master of the data.

Please share your thoughts and objections.  Also, if anyone has ideas about the best way to fit this into the code base (hook, helper, item module, etc.) that would be good too.  I want to do this as unobtrusively as possible.

**Comment by SteveIreland** [comment]
I gave a bit of thought to this awhile back and decided that it should be a two step process.

First, if OSPOS needs to have an interface to a third party package (and it could be numerous popular and obscure packages) then we just set an configuration to "Enable Third Party Interface".

This option would then write the transactions to a common file keyed by a timestamp and identified by a transaction type code.  Then based on the transaction type the body of the transaction would include JSON encoded data needed by third party packages.  Then whoever needs to write an interface could periodically pick up the data in the transaction file and write it to the external package (in the format required by the external package).

Additional complexities could be configured (such as capturing only certain types of transactions) and purging the transaction file of transactions older than what would be of interest.

But that was my thought.

**Comment by jekkos** [comment]
Hmm if you want a reactive flow and have another application react to events from OSPOS then indeed it makes sense to hook another application's API call inside of OSPOS. But that will of course only work for one integration and so it would still be nice to abstract away the way events are generated.

Of course as @SteveIreland suggests it could be done through a dump but some bookkeeping will need to be done at the other side to see up to which point data was imported. However it could be rather straightforward there. This choice doesn't have the benefit that you can't react to events immediately and thus you'll have to poll a REST interface on OSPOS to feed in the data

Another approach might be using reactive data streaming like eg Apache Kafka. So in that case you will need to hook into OSPOS code at certain points (I would say the point where create/update/delete) is done at db level, and then publish these changes to a topic to which other clients are subscribed.

Might be harder at first but it will sort of decouple what needs to be done on the other side from the OSPOS code itself

**Comment by daN4cat** [comment]
@SteveIreland suggested an homemade stream approach instead of Kafka. But adding a queue in between looks elegant but adds complexity for maybe something simple in terms of interfacing an external system via a connector. Publish and subscriber patterns are useful in case of multiple listener. Do we foresee the case?

**Comment by objecttothis** [comment]
I agree that designing this in a way to make multiple connectors usable by the vast majority of other connectors without needing to insert hooks for each one is a better long-term approach.  On the other hand, my requirements include live updates to prevent the possibility of selling the same product twice which could happen if the product sale is not updated on the website from OSPOS until later.  I'd also like to avoid Cron to do that since it introduces issues requiring more knowledge to setup the software.

Using Apache Kafka appeals to me since it resolves the problem of one data center for multiple API's to use.  It's definitely more work, but in my mind worth it.  If we setup OSPOS to generate the kafka data, does it make sense to eventually rewrite SMS (messaging) and mailchimp integrations to utilize it in the same way that the CCIDesq API connecter will use it?

We would then need to figure out the best way to call the "Kafka Consumer" from OSPOS to best reduce the latency.  I'm guessing a post-controller hook is best, but you guys are more experienced in this than I am, so let me know what you think.

**Comment by objecttothis** [comment]
I agree that designing this in a way to make multiple connectors usable by the vast majority of other connectors without needing to insert hooks for each one is a better long-term approach.  On the other hand, my requirements include live updates to prevent the possibility of selling the same product twice which could happen if the product sale is not updated on the website from OSPOS until later.  I'd also like to avoid Cron to do that since it introduces issues requiring more knowledge to setup the software.

Using Apache Kafka appeals to me since it resolves the problem of one data center for multiple API's to use.  It's definitely more work, but in my mind worth it.  If we setup OSPOS to generate the kafka data, does it make sense to eventually rewrite SMS (messaging) and mailchimp integrations to utilize it in the same way that the CCIDesq API connecter will use it?

We would then need to figure out the best way to call the "Kafka Consumer" from OSPOS to best reduce the latency.  I'm guessing a post-controller hook is best, but you guys are more experienced in this than I am, so let me know what you think.

**Comment by jekkos** [comment]
Indeed using post controller hooks or aspects is probably a clean way to separate this integration code from our actual business logic. To make this work I think we need to stick to some common interface definitions for most entities so that they can be hooked into in a generic way. 

I know there were already quite some requests for integrating different sorts of packages and then it's just a matter of tapping into the topic of choice and off you go. We could provide consumers for popular software packages that then listen to changes from OSPOS and do whatever they need to do on the other side.

But @daN4cat you're right that it might add some complexity in terms of setup for sure. A POC on this part would be needed using eg docker to see some simple example working first.

**Comment by objecttothis** [comment]
Based on the above conversation I am writing a post_controller hook in a file called connectors.php which can be used to call all external data slinging connectors.  Right now the only connector that I will write the code for is the CCIDesq connector but it seems like this hook would also be a good place to migrate the mail chimp code to at some point and then when other people want to integrate other external apps they can just write the code in the hook.

**Comment by daN4cat** [comment]
I'm not sure you can make Mailchimp work as post_controller hook if I got right your design intent. Mailchimp requires direct interaction with customer controller business logic and views / forms. It's not a post action type of interaction.

**Comment by erastus** [comment]
@objecttothis  have you solved it?

**Comment by objecttothis** [comment]
@erastus not yet, but it is a priority for me.  I need to see item attributes and import items through first since the API will be dependent on that working correctly.

**Comment by daN4cat** [comment]
@objecttothis what's the difference betwee CCIDesq CMS and CLCdesq?

**Comment by objecttothis** [comment]
> @objecttothis what's the difference between CCIDesq CMS and CLCdesq?

The project was forked into CLCdesq and that is what the API integration will be written for.  Practically there is no difference though.

**Comment by objecttothis** [comment]
@daN4cat @jekkos I'm running into significant problems with using a post_controller hook to create the framework for a messaging service for 3rd party integrations to use the `$ci->db->queries` to monitor and kick off appropriate API calls.  1st, `$ci->db->queries` just gives me strings of the SQL queries so even if I parse them, it doesn't give me the information I need.  For example when an item gets inserted into OSPOS there is an INSERT INTO `ospos_items` that is generated but it doesn't show me the item_id that gets created for that item and would be needed by most integrations to then lookup in the database more information about that item and do its thing.

In order for the proper information to get passed on to whatever 3rd party integration I believe it is best to create an Integration library and call the appropriate master functions (add, update, remove, et cetera) from the Controller while passing the variables that the function will actually need.  This accomplishes the same end as the controller (confining 3rd party integration code to one place) but gives greater control to when the functions get called and what information is available.

Thoughts?

**Comment by daN4cat** [comment]
The database hook was an idea about how to get something "in between".

The Controller hook you are talking about could a call to an interface implemented a plugin.
It all depends on how many we want call because of more than you you have to add a plugin array concept.

**Comment by objecttothis** [comment]
> The database hook was an idea about how to get something "in between".
> 
> The Controller hook you are talking about could be a call to an interface implemented by a plugin.
> It all depends on how many we want call because if many, you have to add a plugin array concept.
> 
> And register and unregister the plugin.

I don't think we should develop this with plugins in their traditional sense in mind.  What I mean is this:
Typically plugins are developed outside of the scope of the project and operate without the same standard of code review which in turn puts security at risk.  This is especially important if we are exposing database queries to these plugins.

I do think it's important that we have a separate code-space within the project so that you don't have code for 3rd party integrations strewn throughout the code.  This means that code for third party integrations goes through the same code reviews but is confined to a library or controller file.  This insures that we can still control the security of the code.

The problem I was bringing up about the post-controller hook is that while it does a stellar job at kicking off the class/function listed in hooks.php, it doesn't pass the information needed for the integration.  For example, the CLCdesq API needs to send information about the item to the CLCdesq API server but all it has to go from when a new item is created is the INSERT INTO SQL statement, but that SQL does not contain the item_id that gets generated and passed back to the item model.  Unless I'm missing something, there is no way for me to use just the db queries to get this kind of information.  This is why I'm thinking that I should use a library in which the class has a set of public functions (integration_add, integration_remove, integration_update).  These functions kick off the private functions in the class appropriate to the API.

**Comment by objecttothis** [comment]
This thread on the CI github page shows someone asking a similar thing about passing data with hooks (https://github.com/bcit-ci/CodeIgniter/issues/4356).  Apparently using hooks as an event dispatcher is not a thing and there is no native event dispatcher.  So it seems like we either implement something like this (https://github.com/nathanmac/CodeIgniter-Events) or I write my own using an integrations library and confine 3rd party integrations code to that library.  Thoughts?

**Comment by jekkos** [comment]
Maybe we can do something generic in a base controller that indeed publishes an event on any crud operation. The interface will look the same for every model and events could be published on a event bus eg which others can then subscribe to.
